osu.Game/Online/API/Requests/PaginatedAPIRequest.cs

```diff
@@ -9,20 +9,27 @@ namespace osu.Game.Online.API.Requests
     public abstract class PaginatedAPIRequest<T> : APIRequest<T> where T : class
     {
         private readonly int page;
+        private readonly int initialItems;
         private readonly int itemsPerPage;
 
-        protected PaginatedAPIRequest(int page, int itemsPerPage)
+        protected PaginatedAPIRequest(int page, int itemsPerPage, int initialItems)
         {
             this.page = page;
+            this.initialItems = initialItems;
             this.itemsPerPage = itemsPerPage;
         }
 
         protected override WebRequest CreateWebRequest()
         {
             var req = base.CreateWebRequest();
 
-            req.AddParameter("offset", (page * itemsPerPage).ToString(CultureInfo.InvariantCulture));
-            req.AddParameter("limit", itemsPerPage.ToString(CultureInfo.InvariantCulture));
+            if (page == 0)
+                req.AddParameter("limit", initialItems.ToString(CultureInfo.InvariantCulture));
+            else
+            {
+                req.AddParameter("offset", (initialItems + (page - 1) * itemsPerPage).ToString(CultureInfo.InvariantCulture));
+                req.AddParameter("limit", itemsPerPage.ToString(CultureInfo.InvariantCulture));
+            }
```

Comment:
Yeah I don't know about this. A lot of implementation detail is hidden here, the initial count is separate to subsequent page count, things are passed from virtual properties to request ctors, it's all a bit bleh IMO.

Would like to see a bit of a smarter structure here. Something like:

```csharp
public readonly struct Pagination
{
    int Offset { get; }
    int Limit { get; }

    public Pagination(int limit)
        : this(limit, 0)
    {
    }

    public Pagination(int limit, int offset)
    {
        Limit = limit;
        Offset = offset;
    }

    public Pagination TakeNext(int limit) => new Pagination
    {
        Offset = Offset + Limit,
        Limit = limit
    };
}
```

Then `PaginatedProfileSubsection` can just declare a `GetNextPage(Pagination? previous)` or something, and all of that logic can be stored there, and you can pass all that down to the request and just embed the plain properties into the request.

Reply:
Sounds good, will give it a whirl.

Reply:
Looks to have simplified things quite well.